Update InferTypeInForEachStatement to respect await foreach

Youssef1313 opened the pull request:
Fixes #48064

Will add a test tomorrow.

Youssef1313 commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs:
```diff
@@ -1230,18 +1230,21 @@ private IEnumerable<TypeInferenceInfo> InferTypeInForEachStatement(ForEachStatem
+
+                var enumerableType = forEachStatementSyntax.AwaitKeyword == default
+                    ? this.Compilation.GetSpecialType(SpecialType.System_Collections_Generic_IEnumerable_T)
+                    : this.Compilation.GetTypeByMetadataName(typeof(IAsyncEnumerable<>).FullName);
```
Cannot make a SpecialType for IAsyncEnumerable because it's not in (Co-located core types)[https://github.com/dotnet/roslyn/blob/master/docs/compilers/Co-located%20core%20types.md]

CyrusNajmabadi replied:
have to check enumerableType for null after this.

Youssef1313 replied:
@CyrusNajmabadi Done

Youssef1313 commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs:
```diff
@@ -1230,18 +1230,21 @@ private IEnumerable<TypeInferenceInfo> InferTypeInForEachStatement(ForEachStatem
+
+                var enumerableType = forEachStatementSyntax.AwaitKeyword == default
+                    ? this.Compilation.GetSpecialType(SpecialType.System_Collections_Generic_IEnumerable_T)
+                    : this.Compilation.GetTypeByMetadataName(typeof(IAsyncEnumerable<>).FullName);
```
Can this fail? should I fallback to IEnumerable to failure?

CyrusNajmabadi commented:
Will merge once test added and green.  Thanks!

Youssef1313 commented:
@CyrusNajmabadi The build is green now.

The test was failing due to IAsyncEnumerable not found in the compilation. I updated the expected result to fallback to IEnumerable and added new test with IAsyncEnumerable included (Not sure if I did that in the best possible way).

CyrusNajmabadi commented:
Can you add tests where the inner type parameter is not inferred to be object?  Thanks!

Youssef1313 commented:
@CyrusNajmabadi Test is added and build is green.

CyrusNajmabadi commented:
Great, thanks!
